---
title: "TnSeq -- prepare reference genome"
date: "`r format(Sys.time(), '%d %B, %Y')`"
author: "Michael Jahn"
output:
  html_notebook: 
    theme: spacelab
    toc: yes
---

## Description

This R notebook describes preparation of a reference genome file in `.fasta` format downloaded from NCBI.

## Libraries

```{r, message = FALSE}
library(tidyverse)
library(stringi)
```

## Annotation of mapped reads

As a prerequisite to assign genome-mapped reads to genes (or other coding regions of interest), we have to **prepare a gene table first**. This table serves as input for the script `DesignRandomPool.pl`, see `README.md` for this github repository. This table is nothing more than a slightly trimmed version of the (RefSeq) `*.gff` file that can be downloaded from NCBI genome.

All reference genome tables were already downloaded in `ref/`, and can be imported from there.

```{r, message = FALSE, warning = FALSE}
# get file names
filenames = list.files("../ref/", pattern = ".gff$")

# apply trimming function
df_ref <- lapply(filenames, FUN = function(filename) {
    
    # read tables
    read_tsv(paste0("../ref/", filename), skip = 9, col_names = FALSE) %>%
    
    # remove unnecessary rows and columns
    filter(X3 != "CDS", X3 != "region") %>%
    select(X1, X4, X5, X7, X3, X9) %>%
    
    # set column names
    rename_with(~ c("scaffold", "begin", "end", "strand", "desc", "locus_tag")) %>%
    
    # extract old and new locustag
    mutate(
      old_locus_tag = stri_extract_first(locus_tag, regex = "CCNA_[R0][0-9]*"),
      new_locus_tag = stri_extract_first(locus_tag, regex = "CCNA_[R0][0-9]*")
    ) %>%
    
    # remove old ID column
    select(-locus_tag)
  })

# inspect head of tables
for (l in df_ref) {print(head(l))}
```

----------

Export the modified tables to `ref/` folder.

```{r}
# assign new filenames
names(df_ref) <- gsub(".gff$", "_trimmed.tsv", filenames)

# and save to disk
silent <- lapply(names(df_ref), function(filename) {
  write_tsv(df_ref[[filename]], file = paste0("../ref/", filename))
})
```

After this, the actions outlined in `README.md` can be performed. This is mainly A) mapping reads to the genome and B) generating a read/barcode pool file (summary) from all mapped reads. These steps are performed using the original `perl` scripts from Morgan Price.
